src/librustc_traits/chalk_context.rs

```diff
@@ -97,24 +99,24 @@ impl context::Context for ChalkArenas<'tcx> {
 
     type Parameter = Kind<'tcx>;
 
-    type ProgramClause = ProgramClause<'tcx>;
```

Comment:
Why doesn’t the terminology line up here anymore with chalk?

Reply:
Mmh at some point the terminology diverged, we use `Clause` in rustc for `ProgramClause` in chalk, and `ProgramClause` in rustc for `ProgramClauseImplication` in chalk. I agree we should somehow resolve that conflict.

Reply:
I think the meaning of these words I would **like** to be using:

- "clause" is one of the "base facts" we know to be true
    - in particular, this includes clauses injected via hypothesis, e.g. the clause `A` in `A => G`.
- "program clause" is something that was given in the "input program"
    - in our case, this means the stuff we created by lowering impls and so forth

In particular, program clauses are available globally at all times, whereas clauses may only be available to some goals.